File: src/services/notifier.ts
import { PrismaClient, users } from "@prisma/client";
import { MongoClient, ServerApiVersion } from "mongodb";

import { NotificationService } from "../interfaces/notification-service";
import { BrasilApiService } from "./brasil-api";
import { WeatherService } from "./weather.api";

export class Notifier {
  private datalake: MongoClient;

  constructor(
    private brasilApiService: BrasilApiService,
    private weatherService: WeatherService,
    private notificationService: NotificationService,
    private prisma: PrismaClient
  ) {
    this.datalake = new MongoClient(process.env.MONGODB_URI as string, {
      serverApi: {
        version: ServerApiVersion.v1,
        strict: true,
        deprecationErrors: true,
      }
    });
  }

  async notifyUsers() {
    const users = await this.prisma.users.findMany();

    console.log(`Enviando notificações para ${users.length} usuários`);

    for (const user of users) {
      await this.notifyUser(user);
    }

    console.log(`Notificações enviadas para ${users.length} usuários`);

    this.datalake.close();
  }

  public async notifyUser(user: users) {
    try {
      const lakeVolume = await this.weatherService.getLakeVolume();
      this.validateUser(user);

      const { postalcode } = user;

      console.log(`Buscando dados para o CEP ${postalcode}`);

      if (!postalcode) {
        throw new Error("CEP não encontrado");
      }

      const postalCodeData = await this.brasilApiService.getPostalCode(
        postalcode
      );

      const tagCityNeightborhood =
        `${postalCodeData.city}_${postalCodeData.neighborhood}`.toUpperCase();

      console.log(
        `Buscando dados de chuva para ${tagCityNeightborhood} (${postalCodeData.location.coordinates.latitude}, ${postalCodeData.location.coordinates.longitude})`
      );

      const { rainfall } = await this.weatherService.getRainfall(
        postalCodeData.location.coordinates.latitude,
        postalCodeData.location.coordinates.longitude
      );

      if (rainfall !== 0 && !rainfall) {
        throw new Error("Não foi possível encontrar o índice de chuva");
      }

      console.log(`Enviando notificação para ${user.name}`);

      const message = this.getNotificationMessage(
        user,
        rainfall,
        lakeVolume.volume
      );

      // Use the notification service to send the notification
      await this.notificationService.sendNotification(user, message);

      console.log(`Notificação enviada para ${user.name}`);

      this.datalake.connect()
        .then(() => {
          console.log("Conectado ao MongoDB");
          const db = this.datalake.db("tethys");
          const collection = db.collection("notifications");
          const document = {
            postal_code: postalcode,
            city: postalCodeData.city,
            neighborhood: postalCodeData.neighborhood,
            state: postalCodeData.state,
            rainfall,
            lake_volume: lakeVolume.volume,
            timestamp: new Date(),
          };

          collection.insertOne(document)
            .then(() => console.log("Dados enviados para o MongoDB"))
            .catch((error) => console.log(error.message));
        })
        .catch((error) => console.log(error.message));
    } catch (error: any) {
      console.log(error.message);
    }
  }

  public getNotificationMessage(
    user: users,
    rainfall: number,
    lakeVolume: number
  ) {
    const { name, postalcode } = user;

    let message = `Olá, ${name}!`;

    if (rainfall + lakeVolume > 80) {
      message += ` Atenção, o índice de chuva para o CEP ${postalcode} está em ${rainfall}mm.`;
      message += ` O volume do lago é de ${lakeVolume}%. Por favor, tome cuidado!`;
    } else {
      message += ` O índice de chuva para o CEP ${postalcode} está em ${rainfall}mm.`;
      message += ` O volume do lago é de ${lakeVolume}%. Indice de chuva e volume do lago estão dentro da normalidade.`;
    }

    return message;
  }

  private validateUser(user: users) {
    if (!user) {
      throw new Error("Usuário não encontrado");
    }

    if (!user.postalcode) {
      throw new Error("CEP não encontrado");
    }

    if (!user.phone) {
      throw new Error("Número de telefone não encontrado");
    }
  }
}

File: services/notifier.py
from services.brasil_api import BrasilApi
from services.weather_api import WeatherService
from services.database import DatabaseService
from interfaces.notification_service import NotificationService

from services.mongodb import MongoService

from datetime import datetime

class Notifier:
    mongo = MongoService()

    def __init__(self, brasil_api: BrasilApi, weather_api: WeatherService,
                 notification_service: NotificationService, database_service: DatabaseService):
        self.brasil_api = brasil_api
        self.weather_api = weather_api
        self.notification_service = notification_service
        self.database_service = database_service

    def notify_users(self):
        for user in self.database_service.get_users():
            if user.email is None:
                continue
            self.notify_user(user)

    def notify_user(self, user):
        lake_volume = self.weather_api.get_lake_volume()
        self.validate_user(user)

        print(f"Usuário {user.name} possui CEP cadastrado")
        postal_code = user.postalcode
        postal_code_data = self.brasil_api.get(postal_code)

        print(f"CEP {postal_code} pertence a cidade {postal_code_data['city']} e bairro {postal_code_data['neighborhood']}")
        tag_city_neightborhood = f"{postal_code_data['city']}_{postal_code_data['neighborhood']}".upper()

        rainfall = self.weather_api.get(postal_code_data["location"]["coordinates"]["latitude"], postal_code_data["location"]["coordinates"]["longitude"])
        message = self.get_notification_message(user, rainfall, lake_volume["volume"])

        print(f"Enviando notificação para {user.email} com a mensagem: {message}")
        self.notification_service.send_notification(user, message)

        print(f"Salvando dados no MongoDB")
        self.mongo.set_document('notifications', {
            "email": user.email,
            "postalcode": postal_code,
            "city": postal_code_data["city"],
            "neighborhood": postal_code_data["neighborhood"],
            "state": postal_code_data["state"],
            "rainfall": rainfall,
            "lake_volume": lake_volume["volume"],
            "tag": tag_city_neightborhood,
            "latitude": postal_code_data["location"]["coordinates"]["latitude"],
            "longitude": postal_code_data["location"]["coordinates"]["longitude"],
            "timestamp": datetime.now()
        })

    def validate_user(self, user):
        if user.postalcode is None:
            raise Exception(f"Usuário {user.name} não possui CEP cadastrado")

    def get_notification_message(self, user, rainfall, lake_volume):
        return f"Olá {user.name}, o volume do lago é de {lake_volume} e a previsão de chuva para sua região é de {rainfall}mm"
